[WFLY-16532][Preview] add ability to configure additional scope for authentication request

PrarthonaPaul opened the pull request:
Analysis doc: https://github.com/wildfly/wildfly-proposals/pull/527
WFLY Issue: https://issues.redhat.com/browse/WFLY-16532
Related ELY Issue: https://issues.redhat.com/browse/ELY-2574
Depends on: https://github.com/wildfly-security/wildfly-elytron/pull/1925
Depends on: https://github.com/wildfly/wildfly/pull/17705

____

<--- THIS SECTION IS AUTOMATICALLY GENERATED BY WILDFLY GITHUB BOT. ANY MANUAL CHANGES WILL BE LOST. --->

> Wildfly issue links:
> * [WFLY-19176](https://issues.redhat.com/browse/WFLY-19176)
> * [WFLY-19078](https://issues.redhat.com/browse/WFLY-19078)

<--- END OF WILDFLY GITHUB BOT REPORT --->

More information about the [wildfly-bot[bot]](https://github.com/wildfly/wildfly-github-bot)

Commit:
[WFLY-16532] Added documentation for the elytron-oidc-client scope attribute

fjuma commented:
@PrarthonaPaul I think it would be good to convert this to a draft PR until we have an Elytron release that includes the required changes.

fjuma commented:
@PrarthonaPaul I've merged the latest updates from upstream into my subsystem bump branch. Once you've updated this to make use of that, you should be able to resolve the conflict here.

fjuma commented:
Just a very minor comment but this [commit](https://github.com/wildfly/wildfly/pull/17018/commits/14ba78a342617f94840c6fcfdc764dcfcce90e9a)could be reworded to make it more clear that this is about OIDC, e.g., "[WFLY-16532] Add the ability to configure scopes with elytron-oidc-client"

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -28,10 +33,29 @@ public class ElytronOidcSubsystemParser_3_0 extends ElytronOidcSubsystemParser_2
+    @Override
+    public PersistentResourceXMLDescription getSecureDeploymentParser() {
+
+        return secureDeploymentParserBuilder
```
I think we need to be a bit careful here because the existing `getSecureDeploymentParser()` method takes the `secureDeploymentParserBuilder` and adds additional attributes to it before building it so I think we're missing attributes here.

I think we need to do something like this instead where we copy the existing value and add the new attribute:

```
final PersistentResourceXMLDescription secureDeploymentParser = secureDeploymentParserBuilder
            .addAttribute(SecureDeploymentDefinition.REALM, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(SecureDeploymentDefinition.PROVIDER, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(RESOURCE, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(CLIENT_ID, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(USE_RESOURCE_ROLE_MAPPINGS, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(BEARER_ONLY, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(ENABLE_BASIC_AUTH, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(PUBLIC_CLIENT, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(TURN_OFF_CHANGE_SESSION_ID_ON_LOGIN, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(TOKEN_MINIMUM_TIME_TO_LIVE, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(MIN_TIME_BETWEEN_JWKS_REQUESTS, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(PUBLIC_KEY_CACHE_TTL, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(ADAPTER_STATE_COOKIE_PATH, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
            .addAttribute(SCOPE, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER) // new in 3.0
            .addChild(redirectRewriteRuleParser)
            .addChild(credentialParser)
            .setUseElementsForGroups(true)
            .build();
...

    @Override
    public PersistentResourceXMLDescription getSecureDeploymentParser() {
            return secureDeploymentParser;
    }
```

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemTransformers.java:
```diff
@@ -52,6 +52,7 @@ public void registerTransformers(SubsystemTransformerRegistration registration)
 
     private static void from3(ChainedTransformationDescriptionBuilder chainedBuilder) {
         ResourceTransformationDescriptionBuilder builder = chainedBuilder.createBuilder(VERSION_3_0_0, VERSION_2_0_0);
+        builder.rejectChildResource(PathElement.pathElement(SECURE_SERVER));
```
When transforming to an older version, if the new `scope` attribute is defined for a `secure-deployment`, we need to reject it. Otherwise, if it's not defined, we can just discard it. Take a look here to see how these checks were done when we added a `trusted-virtual-security-domains` attribute to the `security-domain` resource in the Elytron subsystem:

https://github.com/wildfly/wildfly-core/blob/main/elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java#L189-L192

fjuma replied:
We need to do the same for the `secure-server` resource as well, i.e., if the new `scope` attribute is defined for a `secure-server`, we also need to reject it. Otherwise, if it's not defined, we can just discard it.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/OidcConfigService.java:
```diff
@@ -255,6 +256,10 @@ private String getJSON(ModelNode deployment, boolean convertToRealmConfiguration
                 json.get(REALM).set(authServerUrlAndRealm[1]);
                 json.remove(PROVIDER_URL);
             }
+            String scope = json.get(SCOPE).asStringOrNull();
```
This block of code (the `convertToRealmConfiguration` block) is related to `secure-server` configuration specifically so we don't need to add anything here.

Notice this block, https://github.com/wildfly/wildfly/blob/main/elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/OidcConfigService.java#L239, where the JSON values are set using `SecureDeploymentDefinition.ALL_ATTRIBUTES` so once the SCOPE definition has been added to the `SecureDeploymentDefinition` class as mentioned above and once `ALL_ATTRIBUTES` has been updated to include that, this will already get set automatically.

PrarthonaPaul commented:
> @PrarthonaPaul I've merged the latest updates from upstream into my subsystem bump branch. Once you've updated this to make use of that, you should be able to resolve the conflict here.

Hi @fjuma Thanks for letting me know! Do I just need to pull your changes from the bump again to update? 

And with regards to your comment about the commit message, should I also update the [Jira issue ](https://issues.redhat.com/browse/WFLY-16532) title to also be "Add the ability to configure scopes with elytron-oidc-client" as well?

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ProviderAttributeDefinitions.java:
```diff
@@ -231,6 +231,13 @@ class ProviderAttributeDefinitions {
                     .setValidator(new StringLengthValidator(1, Integer.MAX_VALUE, true, true))
                     .build();
 
+    protected static final SimpleAttributeDefinition SCOPE =
```
This can be removed (this class contains attributes for OIDC providers).

fjuma commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -141,7 +141,8 @@ elytron-oidc-client.secure-deployment.socket-timeout-millis=The timeout for the
 elytron-oidc-client.secure-deployment.connection-timeout-millis=The timeout for establishing a connection with the remote host in milliseconds. A timeout value of zero is interpreted as an infinite timeout, and a negative value is interpreted as undefined.
 elytron-oidc-client.secure-deployment.token-signature-algorithm=The token signature algorithm used by the OpenID provider.
-
+elytron-oidc-client.secure-deployment.credential.scope=The scope values are used to request specific or additional permissions for OIDC.
```
This can be removed.

fjuma commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -141,7 +141,8 @@ elytron-oidc-client.secure-deployment.socket-timeout-millis=The timeout for the
 elytron-oidc-client.secure-deployment.token-signature-algorithm=The token signature algorithm used by the OpenID provider.
-
+elytron-oidc-client.secure-deployment.credential.scope=The scope values are used to request specific or additional permissions for OIDC.
+elytron-oidc-client.secure-deployment.scope=The scope values are used to request specific or additional permissions for OIDC.
```
Maybe we could reword this to something like:

A space separated list of scope values that should be used in the OIDC authentication request. Scopes can be used to request that specific sets of information be made available as Claim Values.

fjuma commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -198,9 +199,11 @@ elytron-oidc-client.secure-server.credential.client-key-alias=The client key ali
 elytron-oidc-client.secure-server.credential.token-timeout=The amount of time after which the token expires and can no longer be used to authenticate requests.
 elytron-oidc-client.secure-server.credential.secret=The client secret that was registered with the OpenID provider.
 elytron-oidc-client.secure-server.credential.algorithm=The credential signature algorithm used by the OpenID provider.
+elytron-oidc-client.secure-server.credential.scope=The scope values are used to request specific or additional permissions for OIDC.
```
This can be removed.

fjuma commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -198,9 +199,11 @@ elytron-oidc-client.secure-server.credential.client-key-alias=The client key ali
 elytron-oidc-client.secure-server.redirect-rewrite-rule.add=Add the rewrite rule for the redirect URI. The rewrite rule is an object notation, where the key is a regular expression with which the redirect URI is matched and the value is the replacement String. You can use the '$' character to use for back-reference in the replacement String.
 elytron-oidc-client.secure-server.redirect-rewrite-rule.remove=Remove the rewrite rule for the redirect URI.
 elytron-oidc-client.secure-server.redirect-rewrite-rule.replacement=The replacement String in the rewrite rule for the redirect URI.
+elytron-oidc-client.secure-server.scope=The scope values are used to request specific or additional permissions for OIDC.
```
We should update the analysis document to mention that the scope attribute is also being added to the `secure-server` resource.

fjuma replied:
Could reword this to: A space separated list of scope values that should be used in the OIDC authentication request. Scopes can be used to request that specific sets of information be made available as Claim Values.

fjuma commented on elytron-oidc-client/src/main/resources/schema/wildfly-elytron-oidc-client_3_0.xsd:
```diff
@@ -1189,6 +1189,16 @@
+                    <xs:documentation>
+                        <![CDATA[
+                        The scope value are used to specify the use of OpedID Connect when authenticating.
+                        Additional scope values can be configured here.
```
Could reword to something like:

A space separated list of scope values that should be used in the OIDC authentication request. Scopes can be used to request that specific sets of information be made available as Claim Values.

fjuma commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/OidcTestCase.java:
```diff
@@ -100,6 +100,13 @@ public void testSecureServerWithProvider() throws Exception {
     }
 
+    @Test
+    public void testSecureDeploymentWithScopes() throws Exception {
```
We could also add a test case for secure-server

fjuma commented on elytron-oidc-client/src/test/resources/org/wildfly/extension/elytron/oidc/oidc.xml:
```diff
@@ -121,6 +121,13 @@
+        <public-client>true</public-client>
+        <ssl-required>EXTERNAL</ssl-required>
+        <scope>profile email phone</scope>
+    </secure-deployment>
```
Could add a secure-server too.

fjuma commented on pom.xml:
```diff
@@ -547,7 +547,7 @@
         <version.org.ow2.asm>9.5</version.org.ow2.asm>
         <version.org.reactivestreams>1.0.4</version.org.reactivestreams>
-        <version.org.wildfly.core>21.1.0.Beta1</version.org.wildfly.core>
+        <version.org.wildfly.core>21.1.0.Beta3-SNAPSHOT</version.org.wildfly.core>
```
This should be removed.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/KeycloakConfiguration.java:
```diff
@@ -57,7 +57,7 @@ public enum ClientAppType {
         DIRECT_ACCESS_GRANT_OIDC_CLIENT,
         BEARER_ONLY_CLIENT,
-        CORS_CLIENT
+        CORS_CLIENT,
```
This change can be removed.

fjuma reviewed (changes requested):
Thanks Prarthona! I've added an initial set of comments. Will take a closer look at the tests next.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/resources/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithSingleScope.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "realm" : "WildFly",
+    "resource" : "SingleScopeApp",
```
We should use client-id and provider-url instead of realm, resource, and auth-server-url.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/resources/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithOpenIDScope.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "realm" : "WildFly",
+    "resource" : "OpenIDScopeApp",
```
We should use client-id and provider-url instead of realm, resource, and auth-server-url.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/resources/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithMultipleScopes.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "realm" : "WildFly",
+    "resource" : "MultipleScopeApp",
```
We should use client-id and provider-url instead of realm, resource, and auth-server-url.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/resources/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithInvalidScope.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "realm" : "WildFly",
+  "resource" : "InvalidScopeApp",
```
We should use client-id and provider-url instead of realm, resource, and auth-server-url.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -264,6 +295,58 @@ public void setup(ManagementClient managementClient, String containerId) throws
 
+            operation = createOpNode(SECURE_DEPLOYMENT_ADDRESS + OPENID_SCOPE_APP + ".war", ModelDescriptionConstants.ADD);
+            operation.get("realm").set(TEST_REALM);
+            operation.get("resource").set(OPENID_SCOPE_APP);
```
Same here and below re: realm and resource.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithDeploymentConfigTest.java:
```diff
@@ -398,6 +441,29 @@ public void testCorsRequestWithEnableCorsWithInvalidOrigin() throws Exception {
+    @Test
+    public void testOpenIDScopeValue() throws Exception {
+        deployer.deploy(OPENID_SCOPE_APP);
+        super.testOpenIDScope();
```
Need to make sure we undeploy the application if this is the last test case that's making use of it.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -346,6 +351,34 @@ public void testCorsRequestWithoutEnableCors() throws Exception {
+    @Test
+    @OperateOnDeployment(OPENID_SCOPE_APP)
+    public void testOpenIDScope() throws Exception {
+        loginToApp(OPENID_SCOPE_APP, org.wildfly.test.integration.elytron.oidc.client.KeycloakConfiguration.ALICE, org.wildfly.test.integration.elytron.oidc.client.KeycloakConfiguration.ALICE_PASSWORD, HttpURLConnection.HTTP_OK, SimpleServlet.RESPONSE_BODY);
```
Can these tests check that the scope was set correctly in the authentication request and that appropriate claims were set in the access tokens based on the scope value?

Commit:
[WFLY-16532] Add the ability to configure scopes with elytron-oidc-client

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/KeycloakConfiguration.java:
```diff
@@ -52,6 +52,10 @@ public class KeycloakConfiguration {
     public static final String CHARLIE_PASSWORD = "charlie123+";
     public static final String ALLOWED_ORIGIN = "http://somehost";
+    public static final String ALICE_First_Name = "Alice";
+    public static final String ALICE_LAST_Name = "Smith";
```
For constants, variable names can be all upper case. The two variable names above are mixing lower case and upper case.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/KeycloakConfiguration.java:
```diff
@@ -184,13 +188,16 @@ private static ClientRepresentation createBearerOnlyClient(String clientId) {
     }
 
-    private static UserRepresentation createUser(String username, String password, List<String> realmRoles) {
+    private static UserRepresentation createUser(String username, String password, List<String> realmRoles, String firstName, String lastName, Boolean emailVerified) {
```
Because there are other usages of `createUser(username, password, realmRoles)`, this won't compile. We need to introduce a `createUser(String username, String password, List<String> realmRoles)` variant that calls this new method signature with the extra values set to null.

We should then update the set calls below so that the first name, last name, and email verified values are only set if not null.

PrarthonaPaul replied:
Thanks for pointing this out! 
Do you mean that there are other calls to `createUser` outside of this repo? I tried compiling just to `WildFly` project and that compiled fine locally. 
But if there are calls from other projects, then I can create a new function as you suggested.

fjuma replied:
Oh sorry, I thought I saw some other calls to `createUser` in this file that hadn't been updated but I see now that these were updated to pass in `null` and `false`. So this will compile.

But to avoid needing to update the existing calls where the new values aren't needed, we can introduce a `createUser(String username, String password, List<String> realmRoles)` variant that calls your new variant with `null` and `false` for the two new params.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -146,49 +148,49 @@ public static void checkDockerAvailability() {
     @OperateOnDeployment(PROVIDER_URL_APP)
     public void testWrongPasswordWithProviderUrl() throws Exception {
-        loginToApp(PROVIDER_URL_APP, org.wildfly.test.integration.elytron.oidc.client.KeycloakConfiguration.ALICE, "WRONG_PASSWORD", HttpURLConnection.HTTP_OK, "Invalid username or password");
+        loginToApp(PROVIDER_URL_APP, org.wildfly.test.integration.elytron.oidc.client.KeycloakConfiguration.ALICE, "WRONG_PASSWORD", HttpURLConnection.HTTP_OK, "Invalid username or password", null, false);
```
Instead of updating the existing references, we can introduce a new `loginToApp` method with the same parameters as before that just calls the new method with the additional values set to `null` and `false`.

Let me know if you have questions about that.

fjuma commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/security/common/servlets/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,79 @@
+package org.jboss.as.test.integration.security.common.servlets;
```
Missing copyright header.

fjuma replied:
Since this new file is specifically for OIDC tests, we can add it in the `testsuite/integration/elytron-oidc-client/subsystem` directory instead.

fjuma commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/security/common/servlets/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        writer.write(RESPONSE_BODY);
+        if(scopeClaims.isEmpty()) {
```
Just a minor nit style wise - it's good to be consistent with white spaces. There should be a white space before the `(` here.

fjuma commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/security/common/servlets/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,79 @@
+        if(scopeClaims.isEmpty()) {
+            writer.write("No additional claims were recovered using the scope values.\n" );
+        }
+        else{
```
This should be on the previous line and a space could be added before the `{`.

fjuma commented on testsuite/shared/src/main/java/org/jboss/as/test/integration/security/common/servlets/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,79 @@
+
+        resp.setContentType("text/plain");
+        final PrintWriter writer = resp.getWriter();
+        if (Boolean.parseBoolean(req.getParameter(CREATE_SESSION_PARAM))) {
```
Is this needed?

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -17,9 +17,6 @@
 package org.wildfly.extension.elytron.oidc;
-
-<<<<<<< HEAD
-<<<<<<< HEAD
```
This should be removed.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemTransformers.java:
```diff
@@ -41,10 +46,26 @@ public String getSubsystemName() {
+
+    private static void from3(ChainedTransformationDescriptionBuilder chainedBuilder) {
+        ResourceTransformationDescriptionBuilder builder = chainedBuilder.createBuilder(VERSION_3_0_0, VERSION_2_0_0);
+        builder.rejectChildResource(PathElement.pathElement(SCOPE));
```
This line can be removed.

fjuma commented on testsuite/shared/pom.xml:
```diff
@@ -259,6 +259,10 @@
+        <dependency>
+            <groupId>org.wildfly.security</groupId>
+            <artifactId>wildfly-elytron</artifactId>
+        </dependency>
```
Once the new servlet is moved as mentioned in a previous comment, I think this won't be needed.

fjuma commented on pom.xml:
```diff
@@ -545,7 +545,15 @@
         <version.org.ow2.asm>9.5</version.org.ow2.asm>
         <version.org.reactivestreams>1.0.4</version.org.reactivestreams>
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Looks like something went wrong when resolving the conflict in this file.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/deployment/OidcWithDeploymentConfigTest.java:
```diff
@@ -181,6 +190,44 @@ public static WebArchive createCorsProviderUrlDeployment() {
+    public static WebArchive createSingleScopeDeployment() {
+        return ShrinkWrap.create(WebArchive.class, SINGLE_SCOPE_APP + ".war")
+                .addClasses(SimpleServlet.class)
+                .addClasses(SimpleSecuredServlet.class)
```
I think we can remove `SimpleSecuredServlet` here and below since the scope related applications use the new servlet instead.

PrarthonaPaul replied:
I had to keep it because otherwise, I was getting a 404 error with just the other servlet.

fjuma replied:
I think it should be possible to remove it. You'll likely need to tweak the path that's used to access the servlet and/or the path that's configured for the servlet. 

Take a look at the path that's used for the existing tests when logging into the app and the path that's configured for the existing servlet. I think that should help identify the changes you'll need.

PrarthonaPaul replied:
I will look into this and update the code. 
All other comments are implemented.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -149,11 +154,42 @@ public static WebArchive createCorsProviderUrlDeployment() {
+    public static WebArchive createSingleScopeDeployment() {
+        return ShrinkWrap.create(WebArchive.class, SINGLE_SCOPE_APP + ".war")
+                .addClasses(SimpleServlet.class)
+                .addClasses(SimpleSecuredServlet.class)
```
Same here and below.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -34,4 +49,65 @@ public class ElytronOidcSubsystemParser_3_0 extends ElytronOidcSubsystemParser_2
+            .addAttribute(MIN_TIME_BETWEEN_JWKS_REQUESTS, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(PUBLIC_KEY_CACHE_TTL, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(ADAPTER_STATE_COOKIE_PATH, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(SCOPE, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER)
```
Would be good to add a comment here like the one above.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -28,27 +43,28 @@ public class ElytronOidcSubsystemParser_3_0 extends ElytronOidcSubsystemParser_2
+            .addAttribute(SecureDeploymentDefinition.REALM, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(SecureDeploymentDefinition.PROVIDER, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(RESOURCE, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
+            .addAttribute(CLIENT_ID, SIMPLE_ATTRIBUTE_PARSER, SIMPLE_ATTRIBUTE_MARSHALLER )
```
It's good to avoid the various white space changes made in this file.

PrarthonaPaul replied:
fixed

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -34,4 +34,65 @@ public class ElytronOidcSubsystemParser_3_0 extends ElytronOidcSubsystemParser_2
         return NAMESPACE_3_0;
     }
+    
+    final PersistentResourceXMLDescription secureDeploymentParser = secureDeploymentParserBuilder
```
Looks like the parser changes are now showing up as part of the subsystem bump commit.

PrarthonaPaul replied:
fixed

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/SecureDeploymentDefinition.java:
```diff
@@ -96,6 +96,12 @@ class SecureDeploymentDefinition extends SimpleResourceDefinition {
 
+    protected static final SimpleAttributeDefinition SCOPE =
+            new SimpleAttributeDefinitionBuilder(ElytronOidcDescriptionConstants.SCOPE, ModelType.STRING, true)
+                    .setAllowExpression(true)
```
Please also update the oidc-expressions.xml test file to configure scopes for a secure-server and a secure-deployment using an expression.

PrarthonaPaul replied:
Thanks. I noticed the values for the attributes are different inside `oidc-expressions.xml`. Would I replace the values for the new secure server and secure deployment to be consistent with the existing values inside `oidc-expressions.xml`? (i. e. `{exp.provider:redhat-sso-expressions}` for provider instead of `keycloak`?)
Additionally, there are more attributes defined for each deployment in oidc-expressions compared to oidc.xml. Should I specify those additional attribute values from the new deployments as well?

fjuma replied:
The `oidc-expressions.xml` test file ensures that every attribute that can be specified using an expression can be parsed successfully.

We can just add the scope attribute to one of the existing `secure-deployments` in that file and to one of the existing `secure-servers` in that file and set its value to an expression.

When taking a look at the expressions in that file, we can see they are formatted as follows:

`${exp.name:some_default_value}`

PrarthonaPaul replied:
Alright, I will add scope to an existing one. 
Thanks

PrarthonaPaul replied:
fixed! Added new scopes to `wildfly-console` and `wildfly-management-expressions.war`

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/KeycloakConfiguration.java:
```diff
@@ -52,6 +52,10 @@ public class KeycloakConfiguration {
     public static final String ALLOWED_ORIGIN = "http://somehost";
+    public static final String ALICE_FIRST_NAME = "Alice";
+    public static final String ALICE_LAST_NAME = "Smith";
+    public static final Boolean ALICE_EMAIL_VERIFIED = true;
```
I think `boolean` could be used here.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -346,22 +351,64 @@ public void testCorsRequestWithoutEnableCors() throws Exception {
         loginToApp(username, password, expectedStatusCode, expectedText, true,
                 new URL("http", TestSuiteEnvironment.getHttpAddress(), TestSuiteEnvironment.getHttpPort(),
-                "/" + appName + SimpleSecuredServlet.SERVLET_PATH).toURI());
+                "/" + appName + SimpleSecuredServlet.SERVLET_PATH).toURI(), null, false);
```
With the new method that was added, we shouldn't need to pass the last two params here.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -346,22 +351,64 @@ public void testCorsRequestWithoutEnableCors() throws Exception {
 
     public static void loginToApp(String appName, String username, String password, int expectedStatusCode, String expectedText, URI requestUri) throws Exception {
-        loginToApp(username, password, expectedStatusCode, expectedText, true, requestUri);
+        loginToApp(username, password, expectedStatusCode, expectedText, true, requestUri, null, false);
```
Same here.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -346,22 +351,64 @@ public void testCorsRequestWithoutEnableCors() throws Exception {
-                "/" + appName + SimpleSecuredServlet.SERVLET_PATH).toURI());
+                "/" + appName + SimpleSecuredServlet.SERVLET_PATH).toURI(), null, false);
+    }
+    public static void loginToApp(String username, String password, int expectedStatusCode, String expectedText, boolean loginToKeycloak, URI requestUri ) throws Exception {
```
Just minor, it's good to have a blank line in between methods.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -346,22 +351,64 @@ public void testCorsRequestWithoutEnableCors() throws Exception {
     }
 
-    public static void loginToApp(String username, String password, int expectedStatusCode, String expectedText, boolean loginToKeycloak, URI requestUri) throws Exception {
+        public static void loginToApp(String username, String password, int expectedStatusCode, String expectedText, boolean loginToKeycloak, URI requestUri, String expectedScope, boolean checkInvalidScope) throws Exception {
```
Looks like the indentation is off here.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -381,7 +429,20 @@ public static void loginToApp(String username, String password, int expectedStat
+                    if (expectedScope != null){
+                        assertTrue(context.toString().contains("scope=" + expectedScope));
+
+                        if(expectedScope.contains("profile")){
```
This is another very minor nit, it's good to use a consistent formatting. To be consistent with the rest of the code, it's good to use spacing like this in if statements:


```
if (someCondition) {
    // do some stuff
} else {
   // do something else
}
```

PrarthonaPaul replied:
Thanks for pointing it out. The new if statements I added don't need to do anything for the else case. For consistency, should I do the following: 
```
if (someCondition) {
    // do some stuff
} else {
   // ignore
}
```

fjuma replied:
No need to introduce else statements where they're not needed. Just wanted to show the overall format for future reference.

PrarthonaPaul replied:
Sounds good! I'll check to make sure I am following this format. Thanks

PrarthonaPaul replied:
fixed

fjuma replied:
Thanks, I still seem to see `if(expectedScope.contains("profile")){` here. Maybe you have some local changes that haven't been pushed yet?

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -263,6 +294,102 @@ public void setup(ManagementClient managementClient, String containerId) throws
+            operation.get("secret").set("secret");
+            Utils.applyUpdate(operation, client);
+
+            operation = createOpNode(SECURE_DEPLOYMENT_ADDRESS + OPENID_SCOPE_APP + ".war", ModelDescriptionConstants.ADD);
```
Is this needed? Looks like it's already added above.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -263,6 +294,102 @@ public void setup(ManagementClient managementClient, String containerId) throws
+            operation.get("secret").set("secret");
+            Utils.applyUpdate(operation, client);
+
+            operation = createOpNode(SECURE_DEPLOYMENT_ADDRESS + SINGLE_SCOPE_APP + ".war", ModelDescriptionConstants.ADD);
```
Is this needed? Looks like it's already added above.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -263,6 +294,102 @@ public void setup(ManagementClient managementClient, String containerId) throws
+            operation.get("secret").set("secret");
+            Utils.applyUpdate(operation, client);
+
+            operation = createOpNode(SECURE_DEPLOYMENT_ADDRESS + INVALID_SCOPE_APP + ".war", ModelDescriptionConstants.ADD);
```
Same here.

PrarthonaPaul replied:
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/OidcWithSubsystemConfigTest.java:
```diff
@@ -263,6 +294,102 @@ public void setup(ManagementClient managementClient, String containerId) throws
+            operation.get("secret").set("secret");
+            Utils.applyUpdate(operation, client);
+
+            operation = createOpNode(SECURE_DEPLOYMENT_ADDRESS + MULTIPLE_SCOPE_APP + ".war", ModelDescriptionConstants.ADD);
```
Same here.

PrarthonaPaul replied:
Sorry, I kept both changes when rebasing. 
Fixed in the new commit

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright (c) 2011, Red Hat, Inc., and individual contributors
```
Very minor, since this is a new file, we can update the year to 2023 :)

PrarthonaPaul replied:
fixed

PrarthonaPaul commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -22,10 +22,9 @@
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.*;
```
fixed

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/OidcBaseTest.java:
```diff
@@ -381,7 +431,19 @@ public static void loginToApp(String username, String password, int expectedStat
+                        if(expectedScope.contains("profile")){
+                            assertTrue(responseString.contains("profile: " + KeycloakConfiguration.ALICE_FIRST_NAME + " " + KeycloakConfiguration.ALICE_LAST_NAME));
+                        }
+                        if(expectedScope.contains("email")){
```
Same here.

PrarthonaPaul replied:
Sorry, I think I dropped some changes when rebasing. 
Should be good now. Thanks for catching it.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @author Prarthona Paul
+ */
+@DeclareRoles({ SimpleServletWithScope.ALLOWED_ROLE })
```
Since we're relying on this role being specified in the Keycloak configuration, it would be good to use the `JBOSS_ADMIN_ROLE` role instead from `KeycloakConfiguration` here.

PrarthonaPaul replied:
fixed!

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+        return (OidcSecurityContext) req.getAttribute(OidcSecurityContext.class.getName());
+    }
+
+    private void displayScopeClaims(String scope, List<String> message, OidcSecurityContext oidcSecurityContext) {
```
Just a minor comment, since this method retrieves the scope values from the request instead of actually displaying them, maybe we can rename this to `retrieveScopeClaims` or `getScopeClaims` or something like that?

PrarthonaPaul replied:
changed!

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+
+        List<String> scopeClaims = new ArrayList<String>();
+        if (oidcSecurityContext != null) {
+            String scope = oidcSecurityContext.getToken().getClaimValueAsString("scope");
```
This line could probably be moved to the `displayScopeClaims` method.

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,20 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
 
+=== Scope Configuration
+
+OpenID Connect allows the client to use scope values to request access to specific privileges. The scopes determine if the client will have access to additional resources when sending a request. According to the https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims[documentations], there are a few scope values that OpenID connect allows, such as: 
```
s/privileges/privileges for access tokens.

fjuma replied:
s/The scopes determine if the client will have access to additional resources when sending a request/An access token's scopes determine what resources will be available when using the access token to access an OAuth2 protected endpoint.

fjuma replied:
"According to the https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims[documentations], there are a few scope values that OpenID connect allows, such as:"

Maybe could reword to something like:

The OpenID Connect https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims[specification] defines a few scope values including:

PrarthonaPaul replied:
fixed!

PrarthonaPaul commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,20 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+
+* `phone` : requests access to the phone_number and phone_number_verified Claims. 
+
+Both the subsystem and deployment configurations allow multiple scope values may be specified using a space delimited and case sensitive list. 
```
Fixed!

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
     String getNameSpace() {
         return NAMESPACE_3_0;
     }
+    final PersistentResourceXMLDescription secureDeploymentParser = secureDeploymentParserBuilder
```
added a newline.

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
+    public PersistentResourceXMLDescription getSecureServerParser() {
+        return secureServerParser;
+    }
+    @Override
```
Fixed!

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
+    public PersistentResourceXMLDescription getSecureDeploymentParser() {
+        return secureDeploymentParser;
+    }
+    @Override
```
Fixed!

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
+            .addChild(credentialParser)
+            .setUseElementsForGroups(true)
+            .build();
+    @Override
```
Fixed!

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
+            .addChild(credentialParser)
+            .setUseElementsForGroups(true)
+            .build();
+    final PersistentResourceXMLDescription secureServerParser = secureServerParserBuilder
```
Fixed!

PrarthonaPaul commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -18,20 +18,87 @@
      * The name space used for the {@code subsystem} element
      */
     public static final String NAMESPACE_3_0 = "urn:wildfly:elytron-oidc-client:3.0";
-
```
Should be fixed now :)

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,95 @@
+    /** The serialVersionUID */
+    private static final long serialVersionUID = 1L;
+    public static final String SERVLET_PATH = "/SimpleServletWithScope";
+    public static final String ALLOWED_ROLE = KeycloakConfiguration.JBOSS_ADMIN_ROLE;
```
I think we can use JBOSS_ADMIN_ROLE directly here instead of assigning it to a new constant variable.

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,23 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+
+* `phone` : requests access to the phone_number and phone_number_verified Claims.
+
+It is possible to configure the scope values that would be used in the OIDC authentication request using the `scope` attribute of a `secure-deployment` or a `secure-server` resource in the `elytron-oidc-client` subsystem.
```
Thanks for the updates! Just to make this easier to spot for a reader who's skimming, just wondering if we could add `==== Subsystem Configuration` and `==== Deployment Configuration` headings to separate this sentence and the next one?

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,26 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+
+* `phone` : requests access to the phone_number and phone_number_verified Claims.
+
+=== Subsystem Configuration
```
Since this section and the next one are part of the Scope Configuration section, I think one more `=` is need here and below.

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,26 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+* `phone` : requests access to the phone_number and phone_number_verified Claims.
+
+=== Subsystem Configuration
+It is possible to configure the scope values that would be used in the OIDC authentication request using the `scope` attribute of a `secure-deployment` or a `secure-server` resource in the `elytron-oidc-client` subsystem. The scope values must be added as a space separated list of lowercase ascii values.
```
The last sentence here applies to both the subsystem and deployment configuration so maybe we could move this sentence to just before the two sections.

Is "lowercase ascii" values mentioned in the OIDC spec?

PrarthonaPaul replied:
the Oidc [Spec](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims:~:text=Multiple%20scope%20values%20MAY%20be%20used%20by%20creating%20a%20space%20delimited%2C%20case%20sensitive%20list%20of%20ASCII%20scope%20values.) mentions that the scope values are case-sensitive.
And yes, I can move it up to the general section.

fjuma replied:
Ok thanks for the pointer. Looks like it mentions "case sensitive" rather than lowercase so we can mention that I think.

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,26 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+
+=== Deployment Configuration
+It is also possible to configure the scope values that would be used via the scope attribute in a deployment's `oidc.json` file. The value for scope must be a space separated list of strings.
+Scopes can be used to request that specific sets of information be made available as Claim Values.
```
"The value for scope must be a space separated list of strings.
Scopes can be used to request that specific sets of information be made available as Claim Values."

This description applies to both cases.

PrarthonaPaul replied:
fixed

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,26 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+=== Deployment Configuration
+It is also possible to configure the scope values that would be used via the scope attribute in a deployment's `oidc.json` file. The value for scope must be a space separated list of strings.
+Scopes can be used to request that specific sets of information be made available as Claim Values.
+Note: If both the deployment and subsystem configurations are configured, then the subsystem configuration will override the deployment values.
```
s/configurations are configured/configuration is present

PrarthonaPaul replied:
fixed! 
Thank you for the comments

Skyllarr commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -141,7 +141,7 @@ elytron-oidc-client.secure-deployment.socket-timeout-millis=The timeout for the
 elytron-oidc-client.secure-deployment.connection-timeout-millis=The timeout for establishing a connection with the remote host in milliseconds. A timeout value of zero is interpreted as an infinite timeout, and a negative value is interpreted as undefined.
 elytron-oidc-client.secure-deployment.token-signature-algorithm=The token signature algorithm used by the OpenID provider.
-
+elytron-oidc-client.secure-deployment.scope=A space separated list of scope values that should be used in the OIDC authentication request. Scopes can be used to request that specific sets of information be made available as Claim Values.
```
@PrarthonaPaul Total minor, we can put the the `elytron-oidc-client.secure-deployment.scope` under the `elytron-oidc-client.secure-deployment.token-signature-algorithm` so it is alphabetically ordered

PrarthonaPaul replied:
Sorry for the delay but this is fixed now! Thank you for your review

pferraro commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemParser_3_0.java:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public class ElytronOidcSubsystemParser_3_0 extends ElytronOidcSubsystemParser_2_0 {
```
Versioning is not well described by polymorphism, as evident by the extent of code duplication.
Consider enumerating your supported subsystem namespaces and parameterizing your PersistentResourceXMLDescription accordingly.
e.g.
https://github.com/wildfly/wildfly-core/blob/main/subsystem-test/tests/src/test/java/org/jboss/as/subsystem/test/stability/FooSubsystemSchema.java

This allows you to build your XML description conditionally, based on version and/or stability level, e.g.
https://github.com/wildfly/wildfly/blob/main/clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebSubsystemSchema.java

pferraro commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcExtension.java:
```diff
@@ -62,7 +64,8 @@ public void initialize(ExtensionContext context) {
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, ElytronOidcSubsystemParser_1_0.NAMESPACE_1_0, ELYTRON_OIDC_SUBSYSTEM_PARSER_1_0);
-        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, ElytronOidcSubsystemParser_2_0.NAMESPACE_2_0, CURRENT_PARSER);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, ElytronOidcSubsystemParser_2_0.NAMESPACE_2_0, ELYTRON_OIDC_SUBSYSTEM_PARSER_2_0);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, ElytronOidcSubsystemParser_3_0.NAMESPACE_3_0, CURRENT_PARSER);
```
This schema should not be registered if the process was started with --stability=DEFAULT
Consider refactoring your subsystem schemas/parsers using PersistentSubsystemSchema and leverage this method for parser registration: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/parsing/ExtensionParsingContext.java#L82

pferraro commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/ExpressionsTestCase.java:
```diff
@@ -35,14 +36,20 @@ public ExpressionsTestCase() {
         if (services != null) return;
         String subsystemXml = "oidc-expressions.xml";
-        services = super.createKernelServicesBuilder(new DefaultInitializer()).setSubsystemXmlResource(subsystemXml).build();
+        services = super.createKernelServicesBuilder(new DefaultInitializer(Stability.COMMUNITY)).setSubsystemXmlResource(subsystemXml).build();
```
You should parameterize this test such that it verifies logic for all subsystem namespace versions/stabilities.

PrarthonaPaul replied:
Do we necessarily need to parametrize this for all versions? 
We are parametrizing the subsystem parsing test cases since it was previously testing the parser for each version. However, here, we are testing expression resolution only for the latest version. I believe parametrizing it would require additional XML config files, and 3 tests instead of 1. 
What do you think?

pferraro replied:
I see.  In this case, we would still want to parameterize this test, but limit the parameters to all *current* schema versions, e.g. ElytronOidcSubsystemSchema.CURRENT.values()

pferraro commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/OidcTestCase.java:
```diff
@@ -87,15 +88,34 @@ public void testSecureServerWithProvider() throws Exception {
+        private final Stability stability;
+
+        public DefaultInitializer(Stability stability) {
+            this.stability = stability;
```
```java
super(stability);

pferraro commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/OidcTestCase.java:
```diff
@@ -87,15 +88,34 @@ public void testSecureServerWithProvider() throws Exception {
+
     private static class DefaultInitializer extends AdditionalInitialization {
 
+        private final Stability stability;
```
This property is redundant.

pferraro commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/Subsystem_3_0_ParsingTestCase.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public class Subsystem_3_0_ParsingTestCase extends AbstractSubsystemBaseTest {
```
If you enumerate your subsystem namespaces, you can easily parameterize your parser tests.
e.g.
https://github.com/wildfly/wildfly-core/blob/main/discovery/src/test/java/org/wildfly/extension/discovery/DiscoverySubsystemTestCase.java

PrarthonaPaul replied:
Hello @pferraro 
Your comments related to parametrizating the tests have been addressed here: https://github.com/wildfly/wildfly/pull/17705

Commit:
WFLY-19078 [Preview] - Update the elytron-oidc-client subsystem parser to enumerate schema versions

PrarthonaPaul commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/ExpressionsTestCase.java:
```diff
@@ -15,34 +15,48 @@
+    public static Iterable<ElytronOidcSubsystemSchema> parameters() {
+        return ElytronOidcSubsystemSchema.CURRENT.values();
+    }
+    public ExpressionsTestCase(ElytronOidcSubsystemSchema schema) {
```
Please note that while I don't use schema in this function, I can't drop it because otherwise I get "IllegalArgument wrong number of arguments". 
I think it is due to the parametrized nature of the tests.

fjuma reviewed (changes requested):
@PrarthonaPaul I've just added a comment on the Elytron PR about an alternative approach we could take for handling unsupported attributes in the `oidc.json` deployment descriptor file. I think we can handle that from `OidcActivationProcessor` instead of adding new logic to Elytron to handle unsupported attributes:

https://github.com/wildfly-security/wildfly-elytron/pull/1925#discussion_r1532368580

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+                    + oidcSecurityContext.getToken().getClaimValueAsString("family_name"));
+        }
+        if (scope.contains("email")){
+            message.add("email: " + oidcSecurityContext.getToken().getClaimValueAsString("email_verified") + "\n");
```
Like my comments on the Elytron PR (https://github.com/wildfly-security/wildfly-elytron/pull/1925#discussion_r1532754462), I think we need to update the tests here to also configure the `address` scope and verify that the `address` claim is present in the obtained token because the current claims used here are already present in the token obtained from Keycloak by default.

PrarthonaPaul replied:
I added the changes, but some tests are failing, will need to look more into this.

fjuma commented on testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/subsystem/SimpleServletWithScope.java:
```diff
@@ -0,0 +1,94 @@
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
```
Very minor, the above should be updated now.

PrarthonaPaul replied:
changed to 2024

fjuma replied:
```
/*
 * Copyright The WildFly Authors
 * SPDX-License-Identifier: Apache-2.0
 */
```

PrarthonaPaul replied:
fixed!

fjuma commented on elytron-oidc-client/src/main/resources/org/wildfly/extension/elytron/oidc/LocalDescriptions.properties:
```diff
@@ -206,6 +206,7 @@ elytron-oidc-client.secure-server.credential.algorithm=The credential signature
 elytron-oidc-client.secure-server.redirect-rewrite-rule.add=Add the rewrite rule for the redirect URI. The rewrite rule is an object notation, where the key is a regular expression with which the redirect URI is matched and the value is the replacement String. You can use the '$' character to use for back-reference in the replacement String.
 elytron-oidc-client.secure-server.redirect-rewrite-rule.remove=Remove the rewrite rule for the redirect URI.
 elytron-oidc-client.secure-server.redirect-rewrite-rule.replacement=The replacement String in the rewrite rule for the redirect URI.
+elytron-oidc-client.secure-server.scope=A space separated list of scope values that should be used in the OIDC authentication request. Scopes can be used to request that specific sets of information be made available as Claim Values.
```
@PrarthonaPaul Have you done any manual tests with `secure-server` configuration with `scope` configured? We should probably create an issue for adding a test case for this to the HAL testsuite (https://github.com/hal/berg). That's where other `secure-server` related tests currently are.

PrarthonaPaul replied:
I have done manual tests and it works. 
I have added the issue for the HAL tests here: https://github.com/hal/berg/issues/118

fjuma commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:
```diff
@@ -141,6 +143,33 @@ configuration). If the subsystem detects that the `OIDC` mechanism is indeed req
+NOTE: Since this attribute is a community level feature, the WildFly server's stability level must be set to community
+level or lower to enable it. The stability level must not be changed to default after configuring the
+`scope` attribute. For more information about the stability levels of the WildFly server, please see
+https://docs.wildfly.org/31/Admin_Guide.html#Feature_stability_levels[here].
```
Instead of hardcoding the URL here, you can specify it the way Diana did here:

https://github.com/wildfly/wildfly/pull/13687/commits/435340810b75956671aaa9cc9ca39e8c7bf2839b#diff-ff7d49f604470857104bc4dd96ab9358f3f857e61fa9d8c23990b1c0eb2f58d2R2675-R2676

That way we don't need to keep updating the link as the version number changes.

PrarthonaPaul replied:
fixed!

fjuma commented:
Note that there's also a conflict in this PR that needs to be updated.

PrarthonaPaul commented on elytron-oidc-client/src/test/java/org/wildfly/extension/elytron/oidc/OidcStabilityTestCase.java:
```diff
@@ -0,0 +1,38 @@
+package org.wildfly.extension.elytron.oidc;
```
fixed!

PrarthonaPaul commented:
Hi @fjuma 
Thanks for the review. 
I have changed this to preview 2.0 and have addressed the comments including the one for OidcActivationProcessor changes to handle unsupported scopes. 
The integration tests are currently failing due to the incorrect stability level of the server. I will need to figure out how to set the stability level for integration tests.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/OidcActivationProcessor.java:
```diff
@@ -48,6 +56,23 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
+                InputStream is = deploymentUnit.getAttachment(Attachments.DEPLOYMENT_ROOT).getRoot().getChild("WEB-INF/oidc.json").openStream();
+                String oidcConfigString = readFromInputStream(is);
+                Map<String, SimpleAttributeDefinition> unsupportedAttributesMap = new HashMap<>();
+                unsupportedAttributesMap.put(ElytronOidcDescriptionConstants.SCOPE, SecureDeploymentDefinition.SCOPE);
```
The attribute name is also available from the attribute definition, e.g., `SecureDeploymentDefinition.SCOPE.getName()` could be used.

fjuma replied:
Just wondering, maybe we could create a `List<SimpleAttributeDefinition> nonDefaultAttributes` in `SecureDeploymentDefinition.java`. Then we could iterate over each of these non default attributes here.

PrarthonaPaul replied:
I can make that change. 
Would it be better to create it inside ElytronOidcDescriptionConstants? Because as we would see in the request RFE, the new attributes are added to ProviderAttributeDefinitions. And if in the future additional attributes get added to other children resources, then they can all access `List<SimpleAttributeDefinition> NON_DEFAULT_ATTRIBUTES` from ElytronOidcDescriptionConstants.

PrarthonaPaul replied:
fixed!

fjuma replied:
For shared attributes, we have a variable that is defined in `ProviderAttributeDefinition`  that also gets used by `SecureDeploymentDefintion`. So we'd likely need to introduce a variable for shared non default attributes when we get to that case.

PrarthonaPaul replied:
Sounds good! 
I will keep that in mind for WFLY-17143

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/OidcActivationProcessor.java:
```diff
@@ -48,6 +56,23 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         if (configService.isSecureDeployment(deploymentUnit) && configService.isDeploymentConfigured(deploymentUnit)) {
             addOidcAuthDataAndConfig(phaseContext, configService, webMetaData);
+        } else if (deploymentUnit.getAttachment(Attachments.DEPLOYMENT_ROOT) != null) {
+            try {
```
Would be good to use `try (InputStream is = ...) {` here.

PrarthonaPaul replied:
fixed!

fjuma replied:
I don't see this change made, maybe it wasn't pushed? The reason  to do this is so that the input stream will get closed for you.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemTransformers.java:
```diff
@@ -31,11 +36,21 @@ public void registerTransformers(SubsystemTransformerRegistration registration)
     private static void from2(ChainedTransformationDescriptionBuilder chainedBuilder) {
-        ResourceTransformationDescriptionBuilder builder = chainedBuilder.createBuilder(VERSION_2_0_0, VERSION_1_0_0);
-        builder.rejectChildResource(PathElement.pathElement(SECURE_SERVER));
+        ResourceTransformationDescriptionBuilder builder = chainedBuilder.createBuilder(VERSION_2_0_0.getVersion(), VERSION_1_0_0.getVersion());
```
I don't think we need transformers for preview features.

PrarthonaPaul replied:
commented out the transformer code for now. We will move forward based on discussions here: https://github.com/wildfly/wildfly/pull/17705#discussion_r1521882991

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/_private/ElytronOidcLogger.java:
```diff
@@ -56,4 +57,7 @@ public interface ElytronOidcLogger extends BasicLogger {
     @Message(id = 8, value = "The 'disable-trust-manager' attribute has been set to 'true' so no trust manager will be used when communicating with the OpenID provider over HTTPS. This value should always be set to 'false' in a production environment.")
     void disableTrustManagerSetToTrue();
 
+    @Message(id = 9, value = "Oidc attribute '%s' is not supported")
```
s/is not supported/is not supported with the current stability level

PrarthonaPaul replied:
fixed!

fjuma commented:
Just adding a reminder on this one so we don't forget: the commits in this PR related to the elytron-oidc-client subsystem refactoring and the introduction of the 2.0:preview schema should be updated once https://github.com/wildfly/wildfly/pull/17705 is finalized.

PrarthonaPaul commented on testsuite/integration/manualmode/src/test/java/org/wildfly/test/manual/elytron/oidc/KeycloakConfiguration.java:
```diff
@@ -0,0 +1,166 @@
+        for (Map.Entry<String, ClientAppType> entry : clientApps.entrySet()) {
+            String clientApp = entry.getKey();
+            ClientRepresentation clientRepresentation = createWebAppClient(clientApp, clientSecret, clientHostName, clientPort, clientApp, false);
+//            clientRepresentation.setDefaultClientScopes(List.of("openid"));
+//            clientRepresentation.setOptionalClientScopes(Arrays.asList("profile", "email", "phone"));
```
Adding this would be nice so that default scopes don't get added to the token automatically and we can check that these are configured only when we ask for it. But this makes the tests fail with a 403 error when simulating a buttonclick. 
Elytron testsuite already tests with this, so, could this be something we don't do on the wildfly testsuite.

fjuma replied:
@PrarthonaPaul I think if we could configure this just for the client needed for the scope related tests, then the other tests that rely on some of these being present shouldn't be affected.

PrarthonaPaul replied:
testsuite/integration/manualmode only has scope-related tests. 
So, I think all clients that are created in this function are scope related. And those were the ones that were failing with the 403 on this line: 
https://github.com/wildfly/wildfly/blob/1353b48d82bb25fec85e0bc1f6e4f9ef7452975e/testsuite/integration/manualmode/src/test/java/org/wildfly/test/manual/elytron/oidc/OidcBaseTest.java#L175
I played around with the config and realised those lines were causing that failure.

fjuma replied:
Thanks for the update, @PrarthonaPaul! Would you be able to create a WFLY issue for updating the default scopes that are used for these test cases and link that issue to WFLY-16532?

I think it should be possible to configure this (since we were able to do something similar from the corresponding tests in the Elytron testsuite).

Maybe something went wrong with the client creation or with adding the users to the realm or something like that. Debugging should help us see what's happening.

This doesn't need to block getting this merged though since we've already verified this from the Elytron testsuite so let's get the follow-up issue created to track this.

PrarthonaPaul replied:
https://issues.redhat.com/browse/WFLY-19176

The-Huginn commented:
Apologies for the spam in the description by the wildfly-bot :sweat_smile: . We accidentally had 2 versions deployed, updating one's another actions. Please feel free to remove all the clutter from the description.

fjuma commented on elytron-oidc-client/src/main/java/org/wildfly/extension/elytron/oidc/ElytronOidcSubsystemTransformers.java:
```diff
@@ -29,16 +34,30 @@ public String getSubsystemName() {
     public void registerTransformers(SubsystemTransformerRegistration registration) {
 
         ChainedTransformationDescriptionBuilder chainedBuilder = TransformationDescriptionBuilder.Factory.createChainedSubystemInstance(registration.getCurrentSubsystemVersion());
+        // 3.0.0 (WildFly 32) to 1.0.0 (WildFly 29)
```
s/1.0.0/2.0.0

Commit:
[WFLY-19176][Preview] Update elytron-oidc-client scope related tests to use optional client scopes for multipleClientScpes test

fjuma commented:
Added the labels for 32.x.

PrarthonaPaul commented:
Hi @fjuma 
I have added the new commit from https://github.com/wildfly/wildfly/pull/17705 and CI has been triggered. 
I will update you on this once it has completed running. 
Thanks!

fjuma commented:
@PrarthonaPaul Thanks for the updates in `OidcActivationProcessor`, they look good! Will keep an eye on CI.

fjuma reviewed (approved):
Adding my approval on this one since follow-up issues have been created for the remaining items (these issues have been linked from https://issues.redhat.com/browse/WFLY-16532).

fjuma commented:
Just checked on CI and there's now a related failure in `OidcWithSubsystemConfigTest.testSubsystemOverride`:

```
java.lang.AssertionError: Expected code == OK but got 403 for request=http://[::1]:8080/SubsystemOverrideOidcApp/SimpleSecuredServlet
  at org.junit.Assert.fail(Assert.java:89)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:359)
  at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:333)
  at org.wildfly.test.integration.elytron.oidc.client.subsystem.OidcWithSubsystemConfigTest.testSubsystemOverride(OidcWithSubsystemConfigTest.java:139)
```

I think I see the issue so will push a fix shortly.

fjuma commented:
This PR is now superseded by https://github.com/wildfly/wildfly/pull/17785 which includes an additional commit to fix the `OidcWithSubsystemConfigTest.testSubsystemOverride` test failure.
